### PR TITLE
TSK-1501: TaskTransferrer cleaned and restructured

### PR DESCRIPTION
**Pull request by tge20**
---
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [ ] I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [ ] Commit message format → TSK-XXX: Your commit message.
- [ ] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

**Comment by tge20** [comment]
https://sonarcloud.io/dashboard?branch=TSK-1501&id=tge20_taskana

File: lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java [hunk 1]
```diff
@@ -241,22 +241,21 @@ void should_BulkTransferOnlyValidTasks_When_SomeTasksToTransferCauseExceptions()
...
+    taskIdList.add(null); // InvalidArgument
+    taskIdList.add("TKI:000000000000000000000000000000000099"); // not existing
     taskIdList.add("TKI:100000000000000000000000000000000006"); // already completed
 
     BulkOperationResults<String, TaskanaException> results =
         taskService.transferTasks("WBI:100000000000000000000000000000000006", taskIdList);
     // check for exceptions in bulk
     assertThat(results.containsErrors()).isTrue();
-    assertThat(results.getErrorMap().values()).hasSize(5);
+    assertThat(results.getErrorMap().values()).hasSize(4);
```
Comment (gitgoodjhe) [inline lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java:254]: Why is there 1 error less now?

**Review by gitgoodjhe — CHANGES_REQUESTED** [review]

File: lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java [hunk 2]
```diff
@@ -138,7 +138,7 @@ void should_ThrowException_When_TaskToTransferDoesNotExist() {
 
     ThrowingCallable call =
         () -> {
-          taskService.transfer("Invalid", "WBI:100000000000000000000000000000000006");
+          taskService.transfer("Invalid", "WBI:100000000000000000000000000000000005");
```
[the remark this answers is not in the document]
Comment (tge20) [reply lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java:141]: For the Workbasket previously used the User does not have a READ permission. Before it had not been a problem because there was another order of checks and thus of the Exception thrown. The renewed implementation first checks the Workbasket Preconditions of the target Workbasket and then the Preconditions for a Task to be transferred. It is consistent with our bulkTransfer.

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskTransferrer.java [hunk 3]
```diff
@@ -57,368 +55,304 @@
...
+   * @throws InvalidArgumentException Thrown if the list is null or empty
+   */
+  BulkOperationResults<String, TaskanaException> transferMultipleByKey(
+      List<String> taskIds, String destinationWorkbasketKey, String destinationDomain)
+      throws NotAuthorizedException, WorkbasketNotFoundException, InvalidArgumentException {
+    LOGGER.debug(
+        "entry to transferMultipleByKey(taskIds = {}, destinationWorkbasketKey = {}, domain = {})",
+        taskIds,
+        destinationWorkbasketKey,
+        destinationDomain);
```
[the remark this answers is not in the document]
Comment (tge20) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskTransferrer.java:158]: The check is implicitly done before when retrieving the Workbasket from the DB. If its key or Id is null or empty the Database will not find a corresponding Workbasket and will thus throw a "WorkbasketNotFoundException". Thus the check is redundant.

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskTransferrer.java [hunk 4]
```diff
@@ -57,368 +55,304 @@
...
-    if (taskIdsToBeTransferred == null) {
-      throw new InvalidArgumentException("TaskIds must not be null.");
+    if (destinationWorkbasket.isMarkedForDeletion()) {
+      throw new WorkbasketNotFoundException(
+          destinationWorkbasket.getId(),
+          String.format(WORKBASKET_MARKED_FOR_DELETION, destinationWorkbasket.getId()));
     }
-    BulkOperationResults<String, TaskanaException> bulkLog = new BulkOperationResults<>();
-    List<String> taskIds = new ArrayList<>(taskIdsToBeTransferred);
-    taskService.removeNonExistingTasksFromTaskIdList(taskIds, bulkLog);
```
[the remark this answers is not in the document]
Comment (tge20) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskTransferrer.java:265]: I think this method inside the TaskService is redundant. We check later on which Tasks can be retrieved from the DB and thus do not also need to check whether inside the idList there are some taskIds equal to zero or null. Moreover that's the only use of the method why I would delete it now (I forgot it before).

File: lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java (hunk 1 above)
Comment (tge20) [reply lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java:254]: Before there existed the method "removeNonExistingTasksFromTaskIdList" which put empty task Id's into the bulkLog (because of the property of a Map only once). I thought it would make more sense to not include NULL Ids as well as EMPTY Ids in our bulkLog. That's why the size of the bulkLog shrinked by one.

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskTransferrer.java [hunk 5]
```diff
@@ -57,368 +55,304 @@
...
+      throws NotAuthorizedException, WorkbasketNotFoundException, InvalidArgumentException {
+    LOGGER.debug(
+        "entry to transferMultipleById(taskIds = {}, destinationWorkbasketId = {})",
+        taskIds,
         destinationWorkbasketId);
-    TaskImpl task = null;
-    WorkbasketSummary oldWorkbasketSummary = null;
-    try {
-      taskanaEngine.openConnection();
-      task = (TaskImpl) taskService.getTask(taskId);
```
[the remark this answers is not in the document]
Comment (tge20) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskTransferrer.java:130]: Which null check you mean? I couldn't find one there...

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java [hunk 6]
```diff
@@ -331,14 +331,14 @@ public Task getTask(String id) throws NotAuthorizedException, TaskNotFoundExcept
   public Task transfer(String taskId, String destinationWorkbasketId)
       throws TaskNotFoundException, WorkbasketNotFoundException, NotAuthorizedException,
           InvalidStateException {
-    return taskTransferrer.transfer(taskId, destinationWorkbasketId);
+    return taskTransferrer.transferSingleById(taskId, destinationWorkbasketId);
```
[the remark this answers is not in the document]
Comment (tge20) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskServiceImpl.java]: I changed them all now into an overloaded method with name transfer(...) and the methods followed by them I called "transferSingleTask" and "transferMultipleTasks". What do you think about this?

**Comment by gitgoodjhe** [comment]
duplicate of https://github.com/Taskana/taskana/pull/1599
